### PR TITLE
nixos/xdg-terminal-exec: init module

### PR DESCRIPTION
**Pull request by Cryolitia**
## Description of changes

Add a NixOS module for already existing package xdg-terminal-exec

CC xdg-terminal-exec maintainer: @quantenzitrone

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by quantenzitrone — CHANGES_REQUESTED** [review]

File: nixos/modules/config/xdg/terminal-exec.nix [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        '';
+      };
+
+      config = mkOption {
+        type = with types; attrsOf (listOf str);
+        default = { };
+        description = ''
+          Configuration options for XDG terminal execution utility.
+
+          The keys are the desktop environments, and the values are the list of teminals' [desktop file ID](https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s02.html#desktop-file-id) to try.
```
Comment (quantenzitrone) [inline nixos/modules/config/xdg/terminal-exec.nix:34]:
```suggestion
          The keys are the desktop environments, and the values are a list of terminals' [desktop file IDs](https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s02.html#desktop-file-id) to try.
```

File: nixos/modules/config/xdg/terminal-exec.nix [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  options = {
+
+    xdg.terminal-exec = {
+      enable = mkOption {
+        default = false;
+        type = types.bool;
+        description = ''
+          Whether to enable the proposal for XDG terminal execution utility.
+        '';
+      };
```
Comment (quantenzitrone) [inline nixos/modules/config/xdg/terminal-exec.nix:26]: enable options are normallly done with `lib.mkEnableOption`
but idk about this one

**Review by quantenzitrone — CHANGES_REQUESTED** [review]
man i don't like the descriptions, but i don't really know how to phrase them better.

XDG Terminal Execution is a Spec proposal, while xdg-terminal-exec is a reference implementation of that spec.
So what are we enabling? The spec? The reference implementation? The proposal?
The configration is afaik part of the spec, so are the options for the utility? For the spec? For any xdg terminal execution utilty implementing the spec?

There are as far as I know no alternative implementations of the spec, but i think it would still make sense to add a package option.

**Comment by Cryolitia** [comment]
> XDG Terminal Execution is a Spec proposal, while xdg-terminal-exec is a reference implementation of that spec. So what are we enabling? The spec? The reference implementation? The proposal? The configration is afaik part of the spec, so are the options for the utility? For the spec? For any xdg terminal execution utilty implementing the spec?
> 
> There are as far as I know no alternative implementations of the spec, but i think it would still make sense to add a package option.

Thanks for pointing out this!

Actually, there's other implementation of it like https://gitlab.com/OroWith2Os/xdg-terminal-exec and https://codeberg.org/mkhl/xdg-terminal-exec , found in https://gitlab.gnome.org/GNOME/glib/-/issues/338

I would be absolutely glad to add the package option in case that if someone someday needs to use other implementation.

To be more clear, would you like to change `pkgs.xdg-terminal-exec` 's description here https://github.com/NixOS/nixpkgs/blob/ac025b8bbb3182b595912516e82fb40fc26c6b29/pkgs/by-name/xd/xdg-terminal-exec/package.nix#L32 to like
> A reference implementation of the proposal for XDG terminal execution utility.

File: nixos/modules/config/xdg/terminal-exec.nix [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+in
+{
+  meta.maintainers = with lib.maintainers; [ Cryolitia ];
+
+  ###### interface
+
+  options = {
+
+    xdg.terminal-exec = {
+      enable = mkEnableOption "Whether to enable the [proposal](https://gitlab.freedesktop.org/xdg/xdg-specs/-/merge_requests/46) for XDG terminal execution utility.";
```
Comment (Aleksanaa) [inline nixos/modules/config/xdg/terminal-exec.nix:20]:
```suggestion
      enable = mkEnableOption "xdg-terminal-exec, the [proposal](https://gitlab.freedesktop.org/xdg/xdg-specs/-/merge_requests/46) for XDG terminal execution utility";
```

File: nixos/modules/config/xdg/terminal-exec.nix [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      package = mkPackageOption pkgs "xdg-terminal-exec" {
+        extraDescription = "The implementation of the XDG terminal execution utility proposal.";
```
Comment (Aleksanaa) [inline nixos/modules/config/xdg/terminal-exec.nix:22]:
```suggestion
```

File: nixos/modules/config/xdg/terminal-exec.nix [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      };
+      config = mkOption {
```
Comment (Aleksanaa) [inline nixos/modules/config/xdg/terminal-exec.nix:24]:
```suggestion
      settings = mkOption {
```

File: nixos/modules/config/xdg/terminal-exec.nix [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        type = with types; attrsOf (listOf str);
+        default = { };
+        description = ''
+          Configuration options for XDG terminal execution utility.
+
+          The keys are the desktop environments, and the values are a list of terminals' [desktop file IDs](https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s02.html#desktop-file-id) to try.
+
+          To define your current desktop environment, run `echo $XDG_CURRENT_DESKTOP`.
+
+          The default value is used when the desktop environment is not found in the configuration.
```
Comment (quantenzitrone) [inline nixos/modules/config/xdg/terminal-exec.nix:32]:
```suggestion
          Configuration options for the Default Terminal Execution Specification.

          The keys are the desktop environments that are matched (case-insensitively) against `$XDG_CURRENT_DESKTOP`,
          or `default` which is used when the current desktop environment is not found in the configuration.
          The values are a list of terminals' [desktop file IDs](https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s02.html#desktop-file-id) to try in order of decreasing priority.
```

I intentionally decreased the amount of whitespace, we don't need everything to be on a new line.

File: nixos/modules/config/xdg/terminal-exec.nix [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        '';
+        example = {
+          default = [ "kitty.desktop" ];
+          GNOME = [ "com.raggesilver.BlackBox.desktop" ];
```
Comment (quantenzitrone) [inline nixos/modules/config/xdg/terminal-exec.nix:36]:
```suggestion
          default = [ "kitty.desktop" ];
          GNOME = [ "com.raggesilver.BlackBox.desktop" "org.gnome.Terminal.desktop" ];
```

an example of multiple terminals is probably good

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 8]
```diff
@@ -149,6 +149,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - [ALVR](https://github.com/alvr-org/alvr), a VR desktop streamer. Available as [programs.alvr](#opt-programs.alvr.enable)
 
+- [xdg-terminal-exec](https://github.com/Vladimir-csp/xdg-terminal-exec), a proposal for XDG terminal execution utility. Available as [xdg.terminal-exec](#opt-xdg.terminal-exec.enable).
```
Comment (quantenzitrone) [inline nixos/doc/manual/release-notes/rl-2405.section.md:152]:
```suggestion
- [xdg-terminal-exec](https://github.com/Vladimir-csp/xdg-terminal-exec), the proposed Default Terminal Execution Specification. Available as [xdg.terminal-exec](#opt-xdg.terminal-exec.enable).
```

**Review by quantenzitrone — CHANGES_REQUESTED** [review]

**Review by quantenzitrone — APPROVED** [review]
tested on my laptop

File: nixos/modules/config/xdg/terminal-exec.nix [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+in
+{
+  meta.maintainers = with lib.maintainers; [ Cryolitia ];
+
+  ###### interface
+
+  options = {
+
+    xdg.terminal-exec = {
+      enable = mkEnableOption "xdg-termonal-exec, the [proposed](https://gitlab.freedesktop.org/xdg/xdg-specs/-/merge_requests/46) Default Terminal Execution Specification";
```
Comment (ambroisie) [inline nixos/modules/config/xdg/terminal-exec.nix:20]:
```suggestion
      enable = mkEnableOption "xdg-terminal-exec, the [proposed](https://gitlab.freedesktop.org/xdg/xdg-specs/-/merge_requests/46) Default Terminal Execution Specification";
```

**Review by ambroisie — DISMISSED** [review]

**Review by LordGrimmauld — APPROVED** [review]
Tested on x86_64-linux.
This works as expected. Interestingly, the launched terminal inherits environment variables of the parent. That makes sense for a lot of things, and i suppose it is within spec. I was just a little surprised to see `xdg-terminal` keeping my environment variables whereas any `xdg-open` that opens a terminal (e.g. nvim on a text file) with `xdg.portal.xdgOpenUsePortal = true;` would reset my env vars to the global session.

I do advise to fix the merge issues though...

File: nixos/modules/config/xdg/terminal-exec.nix [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  cfg = config.xdg.terminal-exec;
+  inherit (lib) mkIf mkEnableOption mkOption mkPackageOption types;
+in
+{
+  meta.maintainers = with lib.maintainers; [ Cryolitia ];
+
+  ###### interface
+
+  options = {
+
```
Comment (drupol) [inline nixos/modules/config/xdg/terminal-exec.nix:18]: Why adding extra empty lines here and not everywhere?

**Review by drupol — APPROVED** [review]
LGTM

**Pull request merged**